volume/driver/overlay_linux.go

```diff
@@ -196,28 +144,149 @@ func (driver *OverlayDriver) workDir(datapath string) string {
-	ancestry := []string{}
+func (driver *OverlayDriver) getGUID(datapath string) string {
+	return filepath.Base(filepath.Dir(datapath))
+}
+
+// NewLiveVolume creates the representation of a volume under the `live`
+// directory.
+//
+// It captures the filepath to the volume, as well as its parents (in a singly
+// linked list).
+//
+// ps.: it DOES NOT protect itself against circular dependencies.
+//
+func NewLiveVolume(root, vol string) (*LiveVolume, error) {
+	volDir := filepath.Join(root, vol)
+	parentSymlink := filepath.Join(volDir, "parent")
+	liveVol := &LiveVolume{Path: volDir}
 
-	currentPath := datapath
-	for {
-		ancestry = append(ancestry, driver.layerDir(currentPath))
+	_, err := os.Stat(volDir)
+	if err != nil {
+		return nil, err
+	}
 
-		parentVolume, err := os.Readlink(filepath.Join(filepath.Dir(currentPath), "parent"))
+	parentDir, err := readlink(parentSymlink)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return liveVol, nil
+		}
+
+		return nil, err
+	}
+
+	liveVol.Parent, err = NewLiveVolume(root, filepath.Base(parentDir))
+	if err != nil {
+		return nil, err
+	}
+
+	return liveVol, nil
+}
+
+// Ancestry traverses LiveVolume linked list producing a list of LiveVolumes
+// from "no dependencies" to "with dependencies".
+//
+func Ancestry(vol LiveVolume) []LiveVolume {
+	res := []LiveVolume{}
+
+	for current := &vol; current != nil; current = current.Parent {
+		res = append([]LiveVolume{*current}, res...)
+	}
+
+	return res
+}
+
+// RecoverMountTable takes care of mounting volumes that exist under `live`, but
+// are not yet mounted due to a system shutdown.
+//
+// It takes care of finding out the dependencies between volumes (represented by
+// symbolic links at `live/<vol>/parent`), and making sure that:
+// - dependencies are mounted first
+// - mounpoints are not mounted more than once
+//
+//  e.g, given the following `live` dir:
+//
+//		.
+//		└── live
+//		    ├── vol1
+//		    ├── vol2
+//		    │   └── parent -> ./vol3
+//		    ├── vol3
+//		    │   └── parent -> ./vol1
+//		    └── vol4
+//			└── parent -> ./vol1
+//
+//
+// it figures out an acceptable mounting order is `vol1`, `vol3`, `vol2`,
+// `vol4`, issuing precisely only 4 `mount`s.
+//
+func RecoverMountTable(root string, mounter Mounter) error {
+	dirs, err := ioutil.ReadDir(root)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+
+		err = fmt.Errorf("readdir %s: %w", root, err)
+		return err
+	}
+
+	mounted := map[string]interface{}{}
+
+	for _, dir := range dirs {
+		vol, err := NewLiveVolume(root, dir.Name())
 		if err != nil {
-			if _, ok := err.(*os.PathError); ok {
-				break
+			return err
+		}
+
+		ancestry := Ancestry(*vol)
+		for idx, ancestor := range ancestry {
+			_, isAlreadyMounted := mounted[ancestor.Path]
+			if isAlreadyMounted {
+				continue
 			}
 
-			return nil, err
-		}
+			datapath := filepath.Join(ancestor.Path, "volume")
+
+			if idx == 0 { // volume with no parents (regular vol)
```

Comment:
we could check the ancestor.Parent == nil to decide

Reply:
[@cirocosta w/ @xtreme-sameer-vohra] : we could actually do some refactoring here to do something like this:

```go


type MountTable struct {
  Mounter Mounter

  storage map[string]interface{}
}

func (m MountTable) isAlreadyMounted(vol LiveVolume) bool {
  _, found := m.storage[vol.Path]
  return found
}


func (m MountTable) markAsMounted(vol LiveVolume) bool {
  _, found := m.storage[vol.Path]
  return found
}

func (m MountTable) Mount(vol LiveVolume) {
  if m.isAlreadyMounted(vol) {
    return
  }

  if vol.Parent == nil {
    m.Mounter.BindMount()
    m.MarkAsMounted(vol)
    return
  }

  m.Mounter.OverlayMount(vol)
  m.MarkAsMounted(vol)
}
```

Reply:
it ends up looking like this:

```go
func RecoverMountTable(root string, mounter Mounter) error {
        dirs, err := ioutil.ReadDir(root)
        if err != nil {
                if os.IsNotExist(err) {
                        return nil
                }

                err = fmt.Errorf("readdir %s: %w", root, err)
                return err
        }

        table := NewMountTable(mounter)

        for _, dir := range dirs {
                vol, err := NewLiveVolume(root, dir.Name())
                if err != nil {
                        return err
                }

                ancestry := Ancestry(*vol)
                for _, ancestor := range ancestry {
                        err = table.Mount(ancestor)
                        if err != nil {
                                return fmt.Errorf("mounting vol %+v: %w", ancestor, err)
                        }
                }
        }

        return nil
}
```

Reply:
clean!